syntax = "proto2";

import "serialization/geometry.proto";
import "serialization/physics.proto";
import "serialization/quantities.proto";

package principia.serialization;

option cc_enable_arenas = true;

message CelestialJacobiKeplerian {
  required int32 celestial_index = 1;
  optional int32 parent_index = 2;
  optional KeplerianElements keplerian_elements = 3;
  required MassiveBody body = 4;
}

message FlightPlan {
  required Quantity initial_mass = 1;
  required Point initial_time = 2;
  required Pair initial_degrees_of_freedom = 12;
  required Point desired_final_time = 3;
  repeated Manoeuvre manoeuvre = 8;
  required AdaptiveStepParameters adaptive_step_parameters = 11;
  optional AdaptiveStepParameters
      generalized_adaptive_step_parameters = 13;  // Added in Erdős.

  // Pre-Cardano.
  reserved 4, 5, 6, 7;
  reserved "integrator",
           "length_integration_tolerance",
           "segment",
           "speed_integration_tolerance";
  // Pre-Буняковский.
  reserved 9, 10;
  reserved "anomalous_segments", "max_steps";
}

message Manoeuvre {
  required Quantity thrust = 1;
  required Quantity initial_mass = 2;
  required Quantity specific_impulse = 3;
  required Multivector direction = 4;
  required Quantity duration = 5;
  required Point initial_time = 6;
  required ReferenceFrame frame = 7;
  // Required, addded in Чебышёв.
  optional bool is_inertially_fixed = 8 [default = true];
}

message Part {
  message PreFrenetInertiaTensor {
    required Quantity mass = 1;
    required SymmetricBilinearForm form = 2;
    required Point centre_of_mass = 3;
  }
  required fixed32 part_id = 4;
  required string name = 9;
  optional bool truthful = 14;  // Added in Frobenius.
  optional Quantity mass = 2;  // Pre-Fréchet, re-added in Frenet.
  optional Point centre_of_mass = 15;  // Added in Galileo.
  optional PreFrenetInertiaTensor
      pre_frenet_inertia_tensor = 10; // Added in Fréchet.
  optional SymmetricBilinearForm inertia_tensor = 12;  // Added in Frenet.
  required Multivector intrinsic_force = 5;
  optional Multivector intrinsic_torque = 13;  // Added in Frenet.
  optional int32 containing_pile_up = 6;
  optional Pair degrees_of_freedom = 1;  // Pre-Fréchet.
  optional RigidMotion rigid_motion = 11;  // Added in Fréchet.
  required DiscreteTrajectory prehistory = 7;
  optional bool tail_is_authoritative = 8;  // Pre-Cesàro.

  // Pre-Cardano.
  reserved 3;
  reserved "gravitational_acceleration_to_be_applied_by_ksp";
  // Pre-Cesàro.
  reserved "tail";
}

message PileUp {
  repeated fixed32 part_id = 1;
  required DiscreteTrajectory history = 4;
  // Pre-Frege.
  map<fixed32, Pair> actual_part_degrees_of_freedom = 5;
  map<fixed32, Pair> apparent_part_degrees_of_freedom = 6;
  // Added in Frege.
  map<fixed32, RigidMotion> actual_part_rigid_motion = 9;
  map<fixed32, RigidMotion> apparent_part_rigid_motion = 10;
  // Added in Frobenius.
  map<fixed32, AffineMap> rigid_pile_up = 11;
  optional EulerSolver euler_solver = 12;
  optional Multivector angular_momentum = 13;  // Required.
  // Added in Cartan.
  optional AdaptiveStepParameters adaptive_step_parameters = 7;
  optional FixedStepParameters fixed_step_parameters = 8;

  // Pre-Cesàro.
  reserved "psychohistory";

  // Pre-Fréchet.
  reserved 2;
  reserved 3;
  reserved "mass";
  reserved "intrinsic_force";
}

message Plugin {
  message CelestialParenthood {
    required int32 index = 1;
    optional int32 parent_index = 2;
    // Added in Catalan.
    optional int32 ephemeris_index = 3;
  }
  message VesselAndProperties {
    required string guid = 1;
    required Vessel vessel = 2;
    required int32 parent_index = 3;
    required bool loaded = 5;
    required bool kept = 7;
    // Pre-Cardano.
    reserved 4;
    reserved "dirty";
  }
  message ZombieAndProperties {
    required string guid = 1;
    required AdaptiveStepParameters prediction_parameters = 2;
  }
  optional fixed64 system_fingerprint = 20;  // Added in Grassmann.
  repeated VesselAndProperties vessel = 1;
  map<fixed32, string> part_id_to_vessel = 16;
  repeated CelestialParenthood celestial = 10;
  required Ephemeris ephemeris = 7;
  required FixedStepParameters history_parameters = 12;
  required AdaptiveStepParameters psychohistory_parameters = 13;
  required Quantity planetarium_rotation = 4;
  required Point game_epoch = 15;
  required Point current_time = 5;
  required int32 sun_index = 6;
  optional ReferenceFrame pre_cauchy_plotting_frame = 11;
  repeated PileUp pile_up = 17;
  optional Renderer renderer = 18;  // Added in Cauchy.
  repeated ZombieAndProperties zombie = 19;  // Added in Frege.
  optional bool uses_correct_sin_cos = 21;  // Added in Lánczos.

  // Pre-Cardano.
  reserved 3;
  reserved "bubble";
  // Pre-Буняковский.
  reserved 8, 9;
  reserved "prolongation_integrator", "prediction_integrator";
  // Pre-Bourbaki.
  reserved 2;
  reserved "/*celestial*/";
  // Pre-Erdős.
  reserved 14;
  reserved "prediction_parameters";
}

// Added in Cauchy.
message Renderer {
  required ReferenceFrame plotting_frame = 1;
}

message Vessel {
  message Checkpoint {
    required Point time = 1;
    required DiscreteTrajectory non_collapsible_segment = 2;
    required FixedStepParameters collapsible_fixed_step_parameters = 3;
  }
  required string guid = 13;
  required string name = 19;
  required MasslessBody body = 1;
  required AdaptiveStepParameters prediction_adaptive_step_parameters = 10;
  repeated Part parts = 14;
  repeated fixed32 kept_parts = 15;
  required DiscreteTrajectory history = 16;
  optional bool psychohistory_is_authoritative = 17;  // Pre-Cesàro.
  optional DiscreteTrajectory prediction = 18;  // Pre-Chasles.
  repeated FlightPlan flight_plans = 4;  // Optional before Hilbert.
  optional int32 selected_flight_plan_index = 24;  // Added in Hilbert.
  optional bool is_collapsible = 22;  // Added in हरीश चंद्र.
  repeated Checkpoint checkpoint = 21;  // Added in हरीश चंद्र.
  optional DiscreteTrajectorySegment.DownsamplingParameters
      downsampling_parameters = 23;  // Added in हरीश चंद्र.

  // Pre-Буняковский.
  reserved 2, 3, 5;
  reserved "/*prediction*/", "history_and_prolongation", "owned_prolongation";
  // Pre-Cardano.
  reserved 6, 7, 8, 9, 11, 12;
  reserved "prolongation_adaptive_step_parameters",
           "history_fixed_step_parameters",
           "prediction_last_time",
           "is_dirty",
           "prediction_fork_time";
  // Pre-Cesàro.
  reserved "psychohistory";
}
